### PR TITLE
fix: DM card fixes, engine improvements, and expanded tests

### PR DESCRIPTION
**Pull request by dmikalova**
- Fix HarmonicPack ability implementation
- Fix Cosmicrux and ExeldonYash abilities
- Fix Bastionclaw
- Add GiltspineMesmerist missing keyword
- Improve PlayCreatureAction and forcedtriggeredabilitywindow engine logic
- Add/expand tests: HarmonicPack, Cosmicrux, Bastionclaw, ExeldonYash, MonsterZero, ShadowGloomcoil, Hydrogan, MemroxTheRed
- Update docs/card-abilities.md and test/AGENTS.md

---

> [!NOTE]
> **Medium Risk**
> Medium risk because it changes core forced-trigger window behavior (auto-resolve rules and dropping invalid queued triggers) and play flow for gigantic creatures, which can affect many cards’ timing/targeting interactions.
> 
> **Overview**
> **Improves forced-trigger resolution and play handling.** The forced triggered ability window now drops queued triggers whose source is no longer in its required `location`, and `autoResolve` only bypasses ordering prompts when *all* simultaneous choices are instances of the same ability.
> 
> **Fixes multiple card implementations.** Updates `Harmonic Pack` to only grant the extra 3 damage when the initial 2 damage was fully applied (while still discarding the revealed archives card when the reveal succeeds), changes `Cosmicrux` to a direct reaction that damages the readied creature, tightens `Exeldon Yash` “if you do” gating to rely on actual `onCapture` events, corrects `Bastionclaw` key-cost modifier to apply to the current player, and marks `Giltspine Mesmerist`’s ready trigger as `autoResolve`.
> 
> **Docs/tests expanded.** Adds documentation for per-ability `autoResolve` and for implementing chained “If you do” clauses via `preThenEvents`, and significantly expands regression tests around forced-trigger ordering, damage/armor/ward/invulnerable edge cases, gigantic deploy/put-into-play, and Memrox archive protections.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 18fc531cdd011ac978254204e007dffb8fe80ede. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: server/game/cards/07-GR/GiltspineMesmerist.js [hunk 1]
```diff
@@ -9,6 +9,7 @@ class GiltspineMesmerist extends Card {
             when: {
                 onCardReadied: (event) => event.exhausted
             },
+            autoResolve: true,
```
Comment (dmikalova) [inline server/game/cards/07-GR/GiltspineMesmerist.js:12]: This is in the same window as cosmicrux, but afaict without other abilities the ordering of this does not matter.

File: server/game/cards/08-AS/HarmonicPack.js [hunk 2]
```diff
@@ -20,34 +20,52 @@ class HarmonicPack extends Card {
...
-                        }))
-                    ]),
-                    message: '{0} uses {1} to {3}discard {4}',
-                    messageArgs: (context) => [
-                        preThenContext.targets.creature ? 'deal 3 more damage and ' : '',
-                        context.preThenEvent.card
-                    ]
-                }
-            })
+            then: (preThenContext) => {
```
Comment (dmikalova) [inline server/game/cards/08-AS/HarmonicPack.js:23]: This came up with exeldon yash - everything before if you do has to be true, otherwise the if you do does not trigger. For HP you must deal 2 damage (after shield/ward/invulnerable) and reveal a card from an archive in order to deal the 3 damage. Then, either way, the revealed card is discarded.

File: server/game/cards/11-VM25/Cosmicrux.js [hunk 3]
```diff
@@ -3,18 +3,13 @@ const Card = require('../../Card.js');
 class Cosmicrux extends Card {
     // After a player readies a creature, deal 1 to it.
     setupCardAbilities(ability) {
-        this.interrupt({
+        this.reaction({
```
Comment (dmikalova) [inline server/game/cards/11-VM25/Cosmicrux.js:6]: Here we go again with Cosmicrux... my understanding now is that while yes all creatures ready at the same time, the way cosmicrux words it causes multiple different orderable 1 damage abilities to trigger (same for giltspine). This means if you have 2 cosmicruxes and a giltspine, then you can interleave the order resolution order of all their 1 damage and discard abilities - if the target dies then the ability still resolves (eg giltspine discard) but if the source dies then it doesn't (eg 1 cosmicrux damage causes the other cosmicrux to be destroyed, then all 1 damage triggers from that second cosmicrux doesn't resolve a la neffru)
[discussion](https://discord.com/channels/563737400275107920/588586239997902858/1502044602692145324)

File: server/game/cards/14-DM/Bastionclaw.js [hunk 4]
```diff
@@ -4,7 +4,7 @@ class Bastionclaw extends Card {
     // Your keys cost –1A for each forged key your opponent has.
     setupCardAbilities(ability) {
         this.persistentEffect({
-            targetController: 'self',
+            targetController: 'current',
```
Comment (dmikalova) [inline server/game/cards/14-DM/Bastionclaw.js:7]: This was reducing enemy key cost

File: server/game/cards/14-DM/ExeldonYash.js [hunk 5]
```diff
@@ -15,12 +15,15 @@ class ExeldonYash extends Card {
                 },
                 then: (preThenContext) => ({
                     condition: () =>
-                        !!preThenContext.target && preThenContext.source.hasToken('amber'),
-                    gameAction: ability.actions.returnAmber((context) => ({
-                        target: context.source,
+                        !!preThenContext.target &&
```
Comment (dmikalova) [inline server/game/cards/14-DM/ExeldonYash.js:18]: See HP comment - for exeldon yash the if you do requires everything before it to happen - you must capture exactly 2 amber and discard a card.

File: server/game/gamesteps/forcedtriggeredabilitywindow.js [hunk 6]
```diff
@@ -82,12 +82,31 @@ class ForcedTriggeredAbilityWindow extends BaseStep {
...
+            return validLoc === sourceLoc;
+        });
+
         if (this.choices.length === 0 || this.pressedDone) {
             return true;
         }
 
-        let autoResolveChoice = this.choices.find((context) => context.ability.autoResolve);
-        if (autoResolveChoice) {
+        const autoResolveChoice = this.choices.find((context) => context.ability.autoResolve);
```
Comment (dmikalova) [inline server/game/gamesteps/forcedtriggeredabilitywindow.js:113]: eg autoresolve multiple giltspines

File: test/server/cards/06-WoE/MemroxTheRed.spec.js [hunk 7]
```diff
@@ -83,4 +83,45 @@ describe('Memrox the Red', function () {
             expect(this.player1.amber).toBe(4);
         });
     });
+
```
Comment (dmikalova) [inline test/server/cards/06-WoE/MemroxTheRed.spec.js:86]: This came up with regards to exeldon > harmonic pack > discarding from archives and I wanted to make sure memrox handled this correctly

File: test/server/cards/14-DM/MonsterZero.spec.js [hunk 8]
```diff
@@ -40,4 +40,116 @@ describe('Monster Zero', function () {
             expect(this.player1).isReadyToTakeAction();
         });
     });
+
```
Comment (dmikalova) [inline test/server/cards/14-DM/MonsterZero.spec.js:43]: More thorough monster zero tests including put into play and deploy

File: test/server/cards/14-DM/ShadowGloomcoil.spec.js [hunk 9]
```diff
@@ -16,5 +18,29 @@ describe('Shadow Gloomcoil', function () {
             expect(this.caspart.damage).toBe(1);
             expect(this.player1).isReadyToTakeAction();
         });
+
```
Comment (dmikalova) [inline test/server/cards/14-DM/ShadowGloomcoil.spec.js:21]: Got a report that gloomcoil was dealing damage after it died in a fight so added that case which seems to be working properly

File: server/game/gamesteps/forcedtriggeredabilitywindow.js [hunk 10]
```diff
@@ -82,12 +82,28 @@ class ForcedTriggeredAbilityWindow extends BaseStep {
             this.deferredChoices = [];
         }
 
+        // Drop queued reactions whose source has left its valid location
```
Comment (dmikalova) [inline server/game/gamesteps/forcedtriggeredabilitywindow.js:92]: If cosmicrux kills itself (or giltspine) then any pending resolutions should be cancelled.
